platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,720 @@
+even if the node is running slower than other nodes in the network. Each node must maintain _at least_ all non-ancient
+events, and should maintain additional non-expired events (though these could be stored on disk to remove them from
+the memory requirement of the node). In addition, _birth round filtering_ limits the number of very old or far future
+events that a node will accept (see [Birth-Round Filtering](#birth-round-filtering)).
+
+#### CPU Pressure
+
+From a high level, either Execution or Consensus can be the primary bottleneck in handling events.
+
+##### Consensus Bottlenecks
+
+Let us suppose that we have a node, Alice. Perhaps initially Alice is able to receive and process events at the same
+speed as other nodes in the network. Perhaps the network load increases past some point that Alice can handle.  At this
+point, other nodes are receiving, creating, gossiping, and coming to consensus on rounds faster than Alice. Remember:
+
+1. Birth-round filtering limits the number of events received by Alice to coincide with the pace at which Alice is
+   handling consensus.
+2. The Tipset algorithm only creates events when doing so will advance consensus. It relies on events that have already
+   passed through Birth-round filtering.
+
+As the other nodes progress farther than Alice, they begin to send events with newer and newer birth rounds. At some
+point, they get too far ahead and begin delivering events that are too far in the future and fail to pass Alice's
+birth-round filter. This prevents Alice from being overwhelmed by events.
+
+Further, since events are coming more slowly to Alice, she will also create her own events more slowly.
+
+Further, since events are coming more slowly to Alice, her Event Creator will update more slowly. When the tipset
+algorithm is consulted for an "other parent", it will find after a short time that there are no other parents it can
+select that will advance consensus. This will cause it to stop creating events until enough events from other creators
+have been processed. This natural slowing of event creation provides a virtuous cycle: as each stressed node slows down
+the event creation rate, it starts to create events with more and more transactions within each event and at a slower
+rate. This will lead to fewer overall events, allowing it to do less work. If the node is still not able to keep up,
+eventually it will refuse to accept any additional transactions from users. If enough nodes in the network are stressed,
+then the overall transaction ingestion rate of the network will be reduced, further reducing the amount of work each
+node has to do. Eventually, an equilibrium is reached.
+
+If the rate at which the network is creating events slows, Alice will be able to catch up by retrieving those previously
+dropped events through gossip, and will be able to process them and catch up. Or, in the last extremity where Alice has
```

Comment:
Removing reference to "dropping events", since it isn't necessary in this context

Reply:
I disagree. Broadcast can send future events. And they will be dropped.

Reply:
In discussions with Leemon, I recall discussing that the broadcast tree protocol will be designed in such a way that far-future rounds are never sent in the first place, by nodes declaring which birth rounds they are able to validate.

@rbair23 I already committed the change here, so let me know if I should revert.

Reply:
> I disagree. Broadcast can send far-future events to a node that is behind.

Although the protocol could in theory be implemented this way, to do so would be super inefficient. The design for broadcast tree chatter has always included sharing of metadata that permits nodes to avoid sending data that would just be discarded by their peer. Additionally, allowing nodes to send data that gets discarded opens attack vectors. If data that is sent may be discarded, then an honest node may be required to send another honest node the same data multiple times. But if an honest node can do this, a dishonest node can simulate it by requesting the same data over and over again. So it's critical to avoid patterns where honest nodes may need to request duplicate transmissions.

Reply:
I think that is a discussion we can have when we design the details of the broadcast gossip algorithm. There are several other criteria that are to be taken into account, including the simplicity and correctness of the implementation. There is not an attack vector here, because no matter what happens, we will have per-node throttles that prohibit malicious nodes from overwhelming neighbors.

Reply:
I've taken the following actions regarding sections that mentioned "dropping" far future events:

- removed
- made intentionally ambiguous, or
- added the caveat that far-future events may just not be sent in the first place

Long story short, Richard had a different understanding of what sending event metadata entailed. After he and I got on the same page, we concluded that broadcast tree gossip will likely not send far-future events in the first place.

I've avoided diving too much into the details on the proposal, though. This will be further explored in future documents.